File: python/bob/bob.py
def is_question(text): return text.endswith("?")
def is_yell(text): return text.isupper()
def is_silence(text): return not text
def is_yell_question(text): return is_yell(text) and is_question(text)
def is_default(text): return True


rules = [
    {"test": is_silence,        "res": "Fine. Be that way!"},
    {"test": is_yell_question,  "res": "Calm down, I know what I'm doing!"},
    {"test": is_yell,           "res": "Whoa, chill out!"},
    {"test": is_question,       "res": "Sure."},
    {"test": is_default,        "res": "Whatever."},
]


def response(hey_bob):
    hey_bob = hey_bob.rstrip()
    for rule in rules:
        if rule["test"](hey_bob):
            return rule["res"]

File: javascript/bob/bob.js
const isSilent = text => text.trim() === '';
const isQuestion = text => text.trim().slice(-1) === "?";
const isShouting = text => /[a-z]/i.test(text) && text.toUpperCase() === text;
const isShoutingQuestion = text => isQuestion(text) && isShouting(text);
const isDefault = () => true;

const rules = [
  {test: isSilent, response: "Fine. Be that way!"},
  {test: isShoutingQuestion, response: "Calm down, I know what I'm doing!"},
  {test: isShouting, response: "Whoa, chill out!"},
  {test: isQuestion, response: "Sure."},
  {test: isDefault, response: "Whatever."},
  
]

export const hey = (message) => {

  return rules.find(rule => rule.test(message)).response;

  // Possible only regex solution
  /*
  switch (true){
    case /^[A-Z\s]+\?$/.test(message):
      return "Calm down, I know what I'm doing!";
    case /^[a-zA-Z0-9,!-@#$%^&*()_+[\]{};:'",\s]*\?\s*$/.test(message):
      return "Sure.";
    case /^[A-Z0-9,!-@#$%^&*()_+[\]{};:'",\s]*[A-Z][A-Z0-9,!-@#$%^&*()_+[\]{};:'",\s]*!*$/.test(message):
      return "Whoa, chill out!";
    case /^\s*$/.test(message):
      return "Fine. Be that way!";
    default:
      return "Whatever.";
  }
  */
};
